inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/base/layer_test_utils.hpp

```diff
@@ -89,27 +97,91 @@ class LayerTestsCommon : public CommonTestUtils::TestsCommon {
+                           << " at index " << i << " with relative threshold " << threshold
                            << " failed";
             }
         }
+*/        
+    }
+
+    template<class T_IE, class T_NGRAPH>
+    static void Compare(const T_NGRAPH *expected, const T_IE *actual, std::size_t size, float threshold, float absThreshold) {
+        if ((threshold < 0) && (absThreshold < 0)) {
+            IE_THROW() << "Both relative threshold and absolute threshold aren't set properly";
+        }
+        std::vector<double> absoluteDifferences;
+        std::vector<double> relativeDifferences;
+        int absoluteErrorCount = 0;
+        int relativeErrorCount = 0;
+        for (std::size_t i = 0; i < size; ++i) {
+            const T_NGRAPH &ref = expected[i];
+            const auto &res = actual[i];
+            const float diff = static_cast<float>(CommonTestUtils::ie_abs(res - ref));
+
+            absoluteDifferences.push_back(diff);
+
+            if (absThreshold >= 0) {
+                if (diff > absThreshold)
+                    absoluteErrorCount++;
+            }
+
+            if (threshold >= 0) {
+                double max;
+                double relDiff;
+                if (sizeof(T_IE) < sizeof(T_NGRAPH)) {
+                    max = std::max(CommonTestUtils::ie_abs(T_NGRAPH(res)), CommonTestUtils::ie_abs(ref));
+                } else {
+                    max = std::max(CommonTestUtils::ie_abs(res), CommonTestUtils::ie_abs(T_IE(ref)));
+                }
+
+                if (max == 0) {
+                    relativeDifferences.push_back(0); // only both res and ref are 0
+                } else {
+                    relDiff = (diff / max);
+                    relativeDifferences.push_back(relDiff);
+                }
+
+                if (relDiff > static_cast<float>(threshold))
+                    relativeErrorCount++;
+
+                if (std::isnan(static_cast<float>(res)) ^ std::isnan(static_cast<float>(ref))) {
+                    IE_THROW() << "One of return values is NaN.  Expected: " << std::to_string(ref) << " and actual: " << std::to_string(res)
+                            << " at index " << i << " failed";
+                }
+            }
+        }
+
+        if ((absoluteErrorCount > 0) || (relativeErrorCount > 0)) {
+            double relDiffMean = std::accumulate(relativeDifferences.begin(), relativeDifferences.end(), 0.0)/relativeDifferences.size();
+            double relDiffMax = *std::max_element(relativeDifferences.begin(), relativeDifferences.end());
+            double absDiffMean = std::accumulate(absoluteDifferences.begin(), absoluteDifferences.end(), 0.0)/absoluteDifferences.size();
+            double absDiffMax = *std::max_element(absoluteDifferences.begin(), absoluteDifferences.end());
+
+            IE_THROW() << "Relative comparison diff tensor mean: " << relDiffMean << ", max: " << relDiffMax << ", # failure(" << relativeErrorCount
```

Comment:
Add the value of thresholds to the error message.
Please make the error messages more formatted using '\t' and '\n' symbols.
And it would be great but not necessary if the output will contain the location of max errors in blob format (loc: {1, 2, 3, 3})

Reply:
Only "\t" doesn't provide pretty output. :) Will check it more and update.
About the location of max error from blob, current Compare() only check 1D vector, need to consider to make variables in LayerTestsCommon class. But several test cases call Compare() directly. Then is is required to show index of max error?

/home/intel-ava14/disk2/openvino_fork/openvino/inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/base/layer_test_utils.hpp:180 
Relative comparison diff tensor mean: 0.00110143,       max: 0.967262,  # failure(1/882) of threshold: 0.01
Absolute comparison diff tensor mean: 1.99932e-06,      max: 9.53674e-06,       # failure(0/882) of threshold: -1

Reply:
![image](https://user-images.githubusercontent.com/1720339/133059045-ed808d27-a970-49cf-b451-e432b5605b1b.png)
Actual output from terminal.

Reply:
Now index is also printed out as below.
Problem is that lots of test cases got failure with relative threshold 0.01(1% accuracy drop).
Do I need to modify all the test cases with higher threshold?

![image](https://user-images.githubusercontent.com/1720339/133177676-24b03207-cbdd-4eae-89e1-801160a0695b.png)

Reply:
/home/jenkins/agent/workspace/private-ci/ie/build-linux-ubuntu20/b/repos/openvino/inference-engine/tests/functional/shared_test_classes/src/base/layer_test_utils.cpp:57

Relative comparison diff tensor mean: 0.0019188, 	max: 1.22222 @ index: 566, 	# failure(2/800) of threshold: 0.01
Absolute comparison diff tensor mean: 1.46493e-06, 	max: 7.62939e-06 @ index: 16, 	# failure(0/800) of threshold: -1

Does this mean that the absolute threshold check is disabled? but relative is enabled?
Enable absolute check by default but a relative check should be manually enabled by the test writer.

Reply:
If absolute threshold = enabled is default, then much more test cases will be failed then current code. And default threshold = 0.01 for absolute threshold seems not to be appropriate for wide range of input value in test cases.
Let me try to measure # of error cases per either absolute default on or relative default on.

Reply:
WIth Absolute comparison is default, failed test case is 1029.
With Relative comparison is default, failed test case is 331.
I used --gtest_filter=smoke*.
How can we make the PR merge ready? Some test case passes only absolute comparison with threshold 0.01 but some others passes relative comparison with threshold 0.01.

<One sample for absolute comparison>
Relative comparison diff tensor mean: 3.46672e-07, 	max: 1.63733e-06 @ index: 1302, 	# failure(0/1920) of threshold: -1
Absolute comparison diff tensor mean: 0.00977071, 	max: 0.046875 @ index: 1302, 	# failure(732/1920) of threshold: 0.01

Reply:
Disable absolute comparison.
For 331 tests manually use only absolute comparison.
For tests that fail again manually set the threshold using `Absolute comparison diff tensor max`( >0.046875 in the example).
Here we will have green tests and can merge.
Then assign Jira tickets to components that failed in tests. Something like `Check absolute threshold int test`

Reply:
Got it! I will work as that and update soon!

Reply:
Resolved the failure cases by using absolute threshold. Will create the tickets for the cases separately.

Reply:
Got more failure cases(3062 test cases) from CI pipeline than local. I think CI covers more test cases than my local(i9-10900k) so I need to debug all of them to pass CI.

Reply:
@apankratovantonp Let me create new PR with the latest OpenVINO base. BTW SubgraphBaseTest::compare() (https://github.com/openvinotoolkit/openvino/blob/e544f5e66fbb7493c09806c21bb0b53f9f9449d9/src/tests/functional/shared_test_classes/src/base/ov_subgraph.cpp#L139) has implemented this kind of statistics error log as you may know. I will implement it into layer_test_util if you agree.